### ISSUE TITLE
[Tiled] Combine TiledMap-Rendering with Entity-Rendering

### ISSUE THREAD

**Issue by McSora** [craftworkgames/MonoGame.Extended#230]
Hey all,

While map-rendering is quite evolving at the moment I'd like to discuss a topic i came across while working the the old tiledmap-renderer.
I'm speaking of a possibillity to integrate entities (character, npcs, dynamically placable structures and so on) in the rendering process of the map.

Of course rendering one map at once is awesome! 
But facing facts, when it comes to oldschool 2D - 4Dir RPGs where you're working with maps with trees or rocks on it which in one case have to be drawn above the player and in another case are drawn beneath the player, the method of rendering single layers is not allways satisfying.

You can think of a tree which is .... 5 squares in height.
A character is about 1.5 squares in height and is colliding 0.5 squares in front of tree.
Let's say we have 2 layers. A bottom layer on which we are placing the bottom of the tree,
and a top layer with the rest of the tree.

We then render as follows:
```
spritebatch.Draw(bottomLayer,...);
spritebatch.Draw(character,...);
spritebatch.Draw(topLayer,...);
```
in this scenario the player would disappear behind the tree just fine.

now think of a character that is taller ... let there be godzilla with a total of 4 squares in height, standing in front of our tree. the bottom of the tree is drawn behind godzilla, the top of the tree in front of him although the tree should disappear completely.
Same goes for a placable structur like a den or something... maybe if you feature some building possibilitires in your game. 
in these cases single-layer-rendering doesn't fit anymore :(

To be honest I dont have a ready-to-use concept for this problem myself...
ie. it would be on the verge of madness to let the map handle a list of entities which it would then ... kind of order by their y-axis ...
I think it could be some higher level class which holds the map and the entities to combine rendering, but now that the map is drawing itself indipendently in the new renderer ... this could also turn out to be tricky.

Maybe it would be best to hear some other voices on the topic :)
Could also be i am thinking in the wrong direction here :D

**Comment by LithiumToast** [comment]
This problem is pretty trivial. Use the Z value to give a depth which position things accordingly. If the depth is related to the height of the object everything will work correctly.

Here is an image explaining the Z value ordering with a scene graph. The image was taken from Apple's documentation [here](https://developer.apple.com/library/ios/documentation/GraphicsAnimation/Conceptual/SpriteKit_PG/Nodes/Nodes.html).
![](https://developer.apple.com/library/ios/documentation/GraphicsAnimation/Conceptual/SpriteKit_PG/Art/node_depth_sorting_2_2x.png)

**Comment by Taikatou** [comment]
How do you give a tile layer a z-value. Is in the tmx file or is it pragmatic?

**Comment by LithiumToast** [comment]
@Taikatou I think the API right now is that in the constructor of the tile layer a depth value can be specified. This is the Z value.

**Comment by Taikatou** [comment]
Hmmm little confused right now I use `Content.Load<TiledMap>(file_name);` and that calls the TiledLayer constructor. Can I load a tilelayer programmatically? :)

**Comment by McSora** [comment]
I'm with Taikatou :O
I am loading the map via Content.Load<TiledMap>(...); which already generates the layers.
And at this moment, these are indipendent of any object that lays outside, like in my case the entities (which of course are not on a layer inside the tmx-file because they can be created dynamically).

But still, if you give one layer a specific depth aren't you running into the same problem i mentioned in my scenario?
I rendered the single layers like
```
spritebatch.Draw(bottomLayer,...);
spritebatch.Draw(character,...);
spritebatch.Draw(topLayer,...);
```

which would, if i am not missing something, do the same thing as 
bottomLayer.Depth = 0
character.Depth = 1
topLayer.Depth = 2
So in this case the problem would remain like descriped before :/
I am very sorry if I am misunderstanding, tho :D

**Comment by craftworkgames** [comment]
It sounds like we are talking about 2 separate issues here:

 1. We need to be able to render *between layers*.
 2. We need to be able to render *between individual tiles*.

I think it's fairly clear that there's 2 clear ways we can render between layers. That's fairly easy.

The part we haven't figured out yet is how to render sprites between the individual tiles. This is particularly important when making an isometric game with a character that moves around like this one:

https://www.youtube.com/watch?v=3DZkYbDhoyo

**Comment by LithiumToast** [comment]
@craftworkgames 

> The part we haven't figured out yet is how to render sprites between the individual tiles.

Each tile would need to be assigned a Z value. It's probably easier to have tiles which have dynamic properties, like their Z values, to be actual sprites instead. For example, in the video you linked all the ground tiles (grass, stone floor, water) would be in one or more layers _always_ below the player. Rocks pillars and other players / NPCs would be sprites since they need to be depth sorted on the fly. The advantage of having some tiles as sprites means that those "sprite tiles" can be added, removed, translated, rotated, etc in game.

**Comment by LithiumToast** [comment]
@Taikatou 

I guess there currently isn't a way to set the `Depth` (Z-value) for layers. Ideally you should be able to do something like this, but the API doesn't support it yet.

```
var map = Content.Load<TiledMap>("...");
// GetLayerByName is currently not in the API, but it would be nice. Could be implemented using a Dictionary<string, TiledLayer>
var groundLayer = map.GetLayerByName<TiledTileLayer>("ground");
groundLayer.Depth = 5; // or whatever you want
groundLayer.Build();
...
foreach (var layer in map.Layers)
{
    layer.Draw(_camera);
}
```

**Comment by McSora** [comment]
@LithiumToast 

I think we will need to be careful with this topic. I can follow your intention and of course NPCs or the character will have to be sprites that will be sorted properly.

But when working with Tiled Maps in Tiled or another tilebased editor, you have got tilesets and
every tileset i worked with so far featuered rocks and trees as a part of the tileset.
Requiring them to be whole sprites would mean everyone had to cut out the trees and rocks and maybe adjust the margin or something for animations or alikes :/
Which would, as i see it, be kind of an antipattern in building tiledbased maps.
You would maybe have to use an image layer or something and place the trees and rocks pixelperfect... :/

As for the Z Values i looked around a bit and 
the RPG Maker for example, they give every tile in the tileset a height value that determines if the tile is above something or below it.
But that would also mean extra effort in tiled .. maybe by adding tile-properties or something...
But at some point one will have to add a specific Z or Height Value to the tiles ... :/

**Comment by LithiumToast** [comment]
@McSora 

> You would maybe have to use an image layer or something and place the trees and rocks pixelperfect... :/

Yup. Which has it's own advantages and disadvantages. Promoting a group of tiles like a house, rock, or tree to a sprite means the sprite can be freely positioned. The advantage is that it's then very easy to encapsulate the sprite in an object that take parts part in the update and draw loop of the game. This would then allow for "tiles" to be given animation, physics, etc. The disadvantage is that it's pain to manage all these objects (including placing them) in Tiled with object groups layers.

> RPG Maker for example, they give every tile in the tileset a height value that determines if the tile is above something or below it.e

In older RPG Makers, if I remember correctly, all the tiles for a layer are given the same Z value. There was ground, mask, and fringe layers. Ground was below the player and had no alpha-blending (opaque). Mask was above the ground, still below the player, but had alpha-blending. Fringe was above the player and had alpha-blending. The trick was to, for example, have the front of a building in a mask layer and the roof of the building in the fringe layer. Then when the player walks in front the of the building the player is in front, but when it walks under a roof tile the player is behind.

> add a specific Z or Height Value to the tiles ... :/

This would also work but it would require the developer to manually assign a Z value to each tile... That's a lot of time wasted on the developer and a nightmare to maintain. If a game had 1000 tiles or more, do you really want to set a Z value manually for each one?

**Comment by McSora** [comment]
@LithiumToast 

> do you really want to set a Z value manually for each one?
short answer? No :laughing: 

And you are right with the advantages and also the disadventages ... 
I myself think, correct me pls if i am wrong,  that we are working on a Tiled api ...
therefore we should avoid setting requirements for developers to do massive overhead work that counteracts with the normal tiled editor workflow...
that includes the cutting of sprites and setting up object groups
and also the setting of Z value i mentioned ^^" ...

> There was ground, mask, and fringe layers.
This again, would mirror the way i rendered sooner using the old map renderer.
i had 2 layers for the ground ... the ground and the mask, they would be rendered first as the bottomlayers
then i rendered the character and the npcs
and last i rendered the toplayer(s).
This system works fine for most cases, but when the character is larger than the object that has a part in the top layer (fringe) but stands in front of it (under it) the bottom part of the rock would be behind the player, the top part would be in front of it ... i hope you understand :D


... is there any case, considering isometric and orthogonal maps... in which "bottom" does not mean "in front of" and "top" does not mean "behind of"?
maybe we could by default link the Z value to the Y value... 
Then we could have still layers like ground and mask, below the npcs
and the fringe above them
and objects like trees would lie on an entities layer and be sorted with other game objects?

**Comment by guFalcon** [comment]
Hi. I was following your discussion and wanted to contribute.
In a past project we've used the sprite-Y and tile-Y as a Z value (sprite-Y/screenHeight to be exact).
We had three map-layers and used the SpriteBatch depth for ordering.
So every layer was drawn in its own SpriteBatch.
First layer was ground. Second layer was obstacles. Then the developer would draw all sprites within the same batch as the second layer. Third layer was roofs, etc...
You could even draw all the layers within a single batch if you'd split the interval right... like
```
ground = [0.0-3.3]
middle = ]3.3-6.6]
top    = ]6.6-1.0]
```
How that would fit into the API is a different question.

**Comment by LithiumToast** [comment]
> I myself think, correct me pls if i am wrong, that we are working on a Tiled api ...

Sure. The Tiled API should allow for setting the depth of a layer and rendering layers individually. Beyond that, it's up to the developer.

> when the character is larger than the object that has a part in the top layer (fringe) but stands in front of it (under it) the bottom part of the rock would be behind the player, the top part would be in front of it

For situations like this it's probably best to have a sprite for the rock. Sprites could be sorted by "Y flip" values using `SpriteBatch`, or the work I done with `PrimitiveBatch` (`ShapeBatch`). A "Y flip" value is a horizontal line for which _other_ sprites with a position below or above the line are rendered in front or behind the sprite respectively. Of course the "Y flip" would be relative to sprite's position.
This approach however is beyond the scope of Tiled renderer.

> ... is there any case, considering isometric and orthogonal maps... in which "bottom" does not mean "in front of" and "top" does not mean "behind of"?

![gl_projectionmatrix02](https://cloud.githubusercontent.com/assets/519592/17488995/c47fca4c-5d50-11e6-9305-430ed568ceb8.png)

In the above image, the figure on the right is the world space. The figure on the left is the orthographic projection of the world space with the camera looking down the -Z axis. This is exactly the coordinate system of `SpriteBatch` except the Z axis is flipped so larger Z values are "behind" and smaller Z values are "in-front".

**Comment by craftworkgames** [comment]
> Sure. The Tiled API should allow for setting the depth of a layer and rendering layers individually. Beyond that, it's up to the developer.

I think the API needs to be flexible enough that people can make the games they want to make without tying ourselves into knots. Obviously we can go too far, but I think we can find a fairly generic solution that allows sprites to be rendered behind tiles.

We absolutely need to support rendering individual layers. That shouldn't be a big deal.

Maybe we can also allow the user to inject a function, or implement an interface that calls a custom draw method during the rendering pass.

Or maybe we can separate the renderer from the tiled map and have a few different kinds of renderers, or allow the renderer to be customized or configured. I know that's a bit vague, I haven't given it a lot of thought yet, but let's think outside the box here.

**Comment by LithiumToast** [comment]
@craftworkgames 

> Obviously we can go too far, but I think we can find a fairly generic solution that allows sprites to be rendered behind tiles.

What we could do is a pass a model-to-world, world-to-view, and view-to-projection matrices to the shader when rendering each tile layer. This is pretty common when rendering any 3D model. See [Displaying a 3D Model on the Screen: Step 4](https://msdn.microsoft.com/en-us/library/bb197293). The model-to-world matrix would be used to allow for a transformation of all the vertices for the tiled layer to be rendered. The result would effectively be specifying a Z value at render time instead of at build time. The world-to-view and view-to-projection matrices can be passed from the `Camera`; I just finished recently improving the `Camera2D` and renaming it to [`OrthographicCamera`](https://github.com/LithiumToast/MonoGame.Extended/blob/3D/Source/MonoGame.Extended/Cameras/OrthographicCamera.cs) in my fork which exposes the view-to-world and view-to-projection matrices as properties.

> maybe we can separate the renderer from the tiled map and have a few different kinds of renderers

Funny you should mention that... I did pretty much that to allow for custom rendering of a `Scene` (new class in my fork). See [here for renderers](https://github.com/LithiumToast/MonoGame.Extended/tree/3D/Source/MonoGame.Extended/Graphics/Rendering). Could easily create a `TiledRenderer` class in the same spirit as what I did with [`SceneRenderer`](https://github.com/LithiumToast/MonoGame.Extended/blob/3D/Source/MonoGame.Extended/Scenes/SceneRenderer.cs) which delegates the rendering completely and of course can be overridden.

**Comment by Taikatou** [comment]
@craftworkgames I really like your idea of separating the tiled_map from the renderer. I think it is a more flexible solution.

**Comment by McSora** [comment]
@Taikatou I agree, I believe it just makes sense to seperate them and @LithiumToast I really like what you did in your fork.
It's kind of the same concept I follow in a project I just started.

Having renderers for each thing that can be rendered.
I got a renderer for the map, a renderer for entities, a renderer for menu and inventory and so on.
And each renderer at least implements the IRenderer interface and at the end there is just a list of all renderers that need to render.

It's funny that I too used the term SceneRenderer ... but that's because i switched to mono from Unity3d .... the things that pissed me off just piled up xD
They say you can use it for 2D purposes but Unity stays a 3D engine... thats another story though...

**Comment by LithiumToast** [comment]
@McSora So, do you think the renderers idea is going to work to solve this issue? Are you okay with having the `Z` or depth of each tile layer "set" when rendering? Sprites/entities would be drawn with their own Z values so the `TiledRenderer` can be isolated from the rest of game rendering logic. The only complication I can think of is that developers would need to come up with a Z value scheme for the order of tile layers / sprites / entities in their game similar to what @guFalcon described. Perhaps there is some sort way to make that simpler and more intuitive with an API.

**Comment by craftworkgames** [comment]
> So, do you think the renderers idea is going to work to solve this issue?

You know what we really need to do. We need to have our demos behave like real games. We should have an isometric tiled maps demo that actually has a character and some enemies walking around on the map. It's ridiculous trying to build this stuff in isolation of making actual games. I mean it kinda worked when we first started but now not so much.

This goes for everything we build in the future. We need a demo, or some way to prove / test that it's a useful feature in a *real* game. So the first thing we need to do to solve this issue is to reproduce the problem by trying to create the demo with the existing API. Once it's clear what the problem is, we can refactor the code to solve it.

**Comment by LithiumToast** [comment]
@craftworkgames 

> Once it's clear what the problem is

The problem is already pretty clear... 

> We should have an isometric tiled maps demo that actually has a character and some enemies walking around on the map. ... This goes for everything we build in the future. We need a demo, or some way to prove / test that it's a useful feature in a real game.

Sure, we should. But I don't see how taking the time to create a demo which incorporates every  problem is a good use of our resources to come up with a _plan_ on how to solve each problem. @guFalcon already proposed a solution to solve the render depth of tile layers and entities, which was used in his real game. The idea I proposed of the "Y-flip" as the Z value is used in this [real game](https://steamcommunity.com/sharedfiles/filedetails/?id=303435958). For isometric games, [ideas expressed on gamedev stack exchange](http://gamedev.stackexchange.com/questions/103442/how-do-i-determine-the-draw-order-of-isometric-2d-objects-occupying-multiple-til) pretty much lean towards promoting tiles to entities.

**Comment by Taikatou** [comment]
I think it would be really great to build this feature alongside a small game. I think what people want is way to store TileMaps and Entitys in a type of World Object. The world can have a Camera2D which decides the draw order.

**Comment by McSora** [comment]
@Taikatou I totally agree with you and if i remember correctly the world object you are talking about would be the Scene object. Unity for example uses the scene as a container for every game object. Okay maybe this is not "exactly" what you are talking about (because the "World" as you called it yould not need to hold some RNG GameObject like it would in unity) but i think this is exactly the direction this would go, a container that holds the map and everything it touches as a plain data-holder.
it holds the map's (or maps') data and all of the entities that would touch the map ... maybe even ambient sound effects, i dont know. 

At the moment the TiledMap seems so ... seperated from the whole other great game functionality extended has to offer (dont feel attacked if i am wrong, i just didnt have the time to go over the whole source :D) I know that its a very young feature and evolving even right now. but we want to support tiledMaps... collision detection... i dont know.. yeah we also want to give every developer the possibility to use sound effects and combine them with objects inside the map.

So, the tiledmap should start holding most generic objects ... call it GameObject.
In the last project i started with a friend (which lead us to Extended btw) we started by creating a gameobject ... for anything our scene object would have to hold.
Going from there, we head VisualGameObjects for rendering (IRenderable so to speak.)

... i even recall seeing that .extended allready has that basic gameobject logic.
i just dont know how far this goes, maybe we allready have objects that are GameObject and implement IRenderable or ICollidable.
If we achieve to combine the tiledmap with those basic objects ... the step towards having a World object that holds map layers ... or tiles as IRenderables ... Sprites or other Objects placed in the TMX file as IRenderables or ICollidables ... and even adding new Objects from outside the map.... which has been created dynamically, wouldn't be too difficult to make anymore.

Sorry for the massive post ... maybe it's a little far fetched for now but as @Taikatou mentioned, this would be something I as a user of the tiledMaps functionality would be really grateful for :D


@craftworkgames Yes i think it would be best to have different game scenarios as demos. I am sure that developers that join in the future would be grateful to see a hand full of samples how to use Extended for different games and even happier if there was exactly the sample they are looking for.

**Comment by craftworkgames** [comment]
> The problem is already pretty clear...

Yes, the problem is clear on paper. That's good. But in my experience, whatever solution we come up with here will have its flaws once we try to use it in the real world. I'm not saying these solutions are bad ideas. They almost certainly *are* the right solutions. As you say, they've already been proven in real games. All I'm saying is that there's only so much we can figure out by discussing it. The rest will be in the details of the implementation. 

> used in this real game. For isometric games, ideas expressed on gamedev stack exchange here and here

I really appreciate references to real games and real answers. Nearly everything we do has already been solved by someone else. There's absolutely no need to try and re-invent the wheel here. Linking to references like these is tremendously helpful to the conversation and gives more credibility to comments. I encourage everyone to do this more often.

> But I don't see how taking the time to create a demo which incorporates every problem is a good use of our resources to come up with a plan on how to solve each problem.

If the only reason to create a demo was to come up with a *plan* I agree it might be a waste of time. However, there's many benefits to creating demos.

 - We have a reusable test case every feature
 - Each demo acts as a form of documentation for people trying to use it
 - We use the feature in a realistic scenario while developing it

That last point is *really* important. Creating "awesome feature X" doesn't really prove anything until someone trying to use it in anger. Every feature looks good as a bullet point on a list, but it's not until you try to use it in a real game that it shows it's true worth. 

I feel so strongly about this that I'm considering making it a requirement of all future PRs. One of the hardest things for me when I take a look at an incoming PR is how the heck I'm going to test it. If I can't test it how do I know if it's a good idea to include it in the library. Half the time I'm not even fully certain what it does.

**Comment by craftworkgames** [comment]
So I've been thinking about this the last couple of days on and off and I think the problem really boils down to 2 options.

Either we allow people to render the individual parts of the map themselves. Something like this:

    _mapBatch.Begin();
    _mapBatch.Draw(_mapLayer0);  // draw the background map layer
    _mapBatch.Draw(_sprite);          // draw custom sprite
    _mapBatch.Draw(_mapLayer1);  // draw the foreground map layer
    _mapBatch.End();

Which kinda works well for rendering things between layers, but gets tricky when trying to render things between tiles.

Or alternately, we allow people to control the depth of the individual parts.

    _mapLayer0.Depth = 0;
    _sprite.Depth = 1;
    _mapLayer1.Depth = 2;

In which case, it doesn't really matter what order things are drawn, because they'll be reordered at the `End` of the batch.

    _mapBatch.Begin();
    _mapBatch.Draw(_map);  // draw the whole map
    _mapBatch.Draw(_sprite);          // draw custom sprite
    _mapBatch.End();

Is this consistent with everyone else's thoughts?

**Comment by LithiumToast** [comment]
@craftworkgames The map doesn't need to be "batched" if making use of the Z, or depth, values for tile layers. It would be more like this

```
map.Draw(camera);
...
spriteBatch.Begin(...);
spriteBatch.Draw(...);
spriteBatch.Draw(...);
spriteBatch.End();
```

> In which case, it doesn't really matter what order things are drawn, because they'll be reordered at the End of the batch.

For sprites using `SpriteBatch` yup. But for the map and things drawn with alpha blending without batching it does matter. See Shawn's [old post on depth sorting](https://blogs.msdn.microsoft.com/shawnhar/2009/02/18/depth-sorting-alpha-blended-objects/). We already ran into this problem with the renderer for the tiled maps; the solution was to render the layers from back to front to fix the problem of unwanted artifacts.

**Comment by craftworkgames** [comment]
Ah right, that makes sense after reading Shawn's depth sorting post. So just to be clear we are talking about either using depth buffering vs painters algorithm. 

When I'm using a `SpriteBatch` I usually think about it in terms of the painters algorithm and I usually think about the boundaries of what I'm drawing between `Begin` and `End`. 

What you're saying if I'm understanding correctly, is that when using depth buffering you can draw many *batches* of things in whatever order and they will be drawn *correctly*. Ignoring alpha-blending of course.

The difficultly then is understanding how to apply depth to the sprites. We haven't exposed anything in the API to make it clear that setting a sprites depth to Y will draw between layers X and Z.

**Comment by LithiumToast** [comment]
@craftworkgames 

> The difficultly then is understanding how to apply depth to the sprites. We haven't exposed anything in the API to make it clear that setting a sprites depth to Y will draw between layers X and Z.

`SpriteBatch.Draw` method has a `depth` parameter. This is the Z value. You can see it being applied [here](https://github.com/MonoGame/MonoGame/blob/develop/MonoGame.Framework/Graphics/SpriteBatchItem.cs#L28).

**Comment by craftworkgames** [comment]
I understand that part. What I'm saying is that there's no way of knowing what *value* to assign to the depth parameter or even that it's required.

We could provide some sort of *documentation* to explain how it works of course. To explain that layers are assigned some arbitrary depth and that each tile within a layer is also assigned some arbitrary depth value. Then explain that sprites drawn at a particular depth will be drawn between tiles A and B on layer C.

Maybe that's the only way we can do it. However, I'm also wondering if documentation is really the best approach here. Maybe it would be better to expose a way to let the user come up with the depth assignment rules. I dunno, just thinking out loud here.

My point is that I didn't understand how it works even after looking at the code and I don't expect many other people to understand it either. The problem we are trying to solve is not really a technical one, it's a human one. How can we design our API so that it's obvious what's going on?

**Comment by craftworkgames** [comment]
Okay so I played around with it briefly and it's currently impossible to get it to work because map layers are being drawn at depths between 1.0 and 5.0 (set by the max number of layers) due to this line:

    _basicEffect.Projection = Matrix.CreateTranslation(projectionHalfPixel) * Matrix.CreateOrthographicOffCenter(
                    left: 0,
                    right: _graphicsDevice.Viewport.Width,
                    bottom: _graphicsDevice.Viewport.Height,
                    top: 0,
                    zNearPlane: 1.0f,
                    zFarPlane: _highestZ);

And sprites are being drawn between 0.0 and -1.0 due to this line in `SpriteBatch.cs` (the last 2 numbers are the near and far plane respectively:

    Matrix.CreateOrthographicOffCenter(0.0f, (float) viewport.Width, (float) viewport.Height, 0.0f, 0.0f, -1f, out result);

So it doesn't matter what depth we set the sprite, it will never be between any layers.

Anyway, I understand the problem a lot better now. I'll have a think about how to proceed and come back to it later.

**Comment by LithiumToast** [comment]
I think the idea of using the painter's algorithm to render the scene from back to front with each tile layer individual like you mentioned earlier is probably going to be one solution. It is intuitive and completely avoids explaining or setting depth values. The responsibility is then on the developer to have a `SpriteBatch.Begin` and `SpriteBatch.End` where required.

e.g.
```
// draw ground tile layer
groundTileLayer.Draw(camera);
// draw mask tile layer
maskTileLayer.Draw(camera);
// draw entities
spriteBatch.Begin(...);
spriteBatch.Draw(...);
spriteBatch.Draw(...);
spriteBatch.End();
// draw fringe tile layer
fringeTileLayer.Draw(camera);
```

Another possible possible approach is to use the `SceneGraph` for depth sorting `SceneNode`s like I mentioned way earlier in this issue. Each `SceneNode` could be a tile layer, or an entity, or whatever. A tile layer `SceneNode` could have children `SceneNode`s which are entities for example. The assignment of depth values would be arbitrary... However it does have the advantage of being extremely easy to manage the depth of tile layers and entites since each `SceneNode` would inherit the depth of it's parent `SceneNode`. The `SceneGraph` could then be walked to have a painter's algorithm with depth buffering.

> And sprites are being drawn between 0.0 and -1.0 due to this line in SpriteBatch.cs (the last 2 numbers are the near and far plane respectively

By default in the right-hand coordinate system, the -Z direction is "looking into the screen" and the +Z direction is "looking out of the screen". i.e. `1.0f` would be infront of `0.0f`. `SpriteBatch` has the opposite API for depth values where `0.0f` is the minimum value for things in front and `1.0f` is the maximum value for things behind. This is why the near plane has a Z value `0.0f` and the far plane has Z value of `-1.0f` in `SpriteBatch`'s orthographic projection.

> So it doesn't matter what depth we set the sprite, it will never be between any layers.

As long as the projection is the same for the tile layers and the tile layers are given a Z value between `0.0f` and `1.0f`, they can be between sprites rendered with `SpriteBatch`.

**Comment by craftworkgames** [comment]
> I think the idea of using the painter's algorithm to render the scene from back to front with each tile layer individual like you mentioned earlier is probably going to be one solution.

The painters algorithm is certainly the most intuitive one for people used to using a `SpriteBatch` but it doesn't help to solve depth issues between tiles on isometric maps.

> As long as the projection is the same for the tile layers and the tile layers are given a Z value between 0.0f and 1.0f, they can be between sprites rendered with SpriteBatch.

Yes exactly, but that's currently not the case.

I'm going to keep thinking about the API and try out a few ideas. Currently I'm thinking we need a few more parameters on the map `Draw` methods to give the user more control of how things are rendered. One of those things could be a *depth function* that takes a layer index and returns a depth value for it.

To be honest I've never really liked passing the camera into the `Draw` method. It should probably just take a transformationMatrix like the SpriteBatch.Begin does.

**Comment by rafaelalmeidatk** [comment]
> We need to be able to render between layers.

This one is easy, we can just use the depth value, right? Well, I tried it a time ago and figured that I calculated the depth of the layers very wrong, I didn't know that the Z value of the `SpriteBatch` only goes between `0.0f` and `1.0f`. Actually, the Z value of the layers is like that:

- Layer 1: `100.0f`
- Layer 2: `200.0f`
- Layer 3: `300.0f`

And so on. I made it like that to let the programmer be able to render between the layers, `100.0f` "of distance" seemed a good number. To solve it is very simple, just insert the layers Z values inside the `0.0f`~`1.0f` range, what can be achieved with a simple division. The problem is that as the range is only between `0.0f` and `1.0f`, we have a very small distance between the layers, this very small numbers (like `0.05f`) numbers doesn't seem so friendly.

**Comment by LithiumToast** [comment]
@craftworkgames 

> but it doesn't help to solve depth issues between tiles on isometric maps.

I honestly think that isometric games should have everything positioned in 3D and then use an orthographic projection to flatten everything to 2D with billboard (camera facing) sprites. Positioning isometric tiles within 3D space is a hell of a lot simpler problem then doing it within 2D space.

> One of those things could be a depth function that takes a layer index and returns a depth value for it.

Why not just use a transform matrix to translate the position of all triangles for the tiles to the correct depth value when rendering? The depth value for each layer could then be in a public get/set property.

> To be honest I've never really liked passing the camera into the Draw method. It should probably just take a transformationMatrix like SpriteBatch.Begin does.

Ah, in my fork I have the camera class calculate the `WorldToView` and the `ViewToProjection` matrices, which are then used for rendering. For example, with `ShapeBatch.Begin` I can pass a `worldToView` matrix (this would be the same as `SpriteBatch`s transform matrix) and a `viewToProjection` matrix as parameters. It's a shame that `SpriteBatch` doesn't support optionally changing the projection matrix... Could solve the problem of changing the near and far plane to have Z values beside `0.0f` and `1.0f`.

**Comment by LithiumToast** [comment]
@rafaelalmeidatk 

> The problem is that as the range is only between 0.0f and 1.0f, we have a very small distance between the layers, this very small numbers (like 0.05f) numbers doesn't seem so friendly.

I think that's a big problem why the depth values for `SpriteBatch` are unintuitive. I think an API that translates a friendly depth value into a depth value for rendering could go a long ways.

**Comment by craftworkgames** [comment]
> The map doesn't need to be "batched" if making use of the Z, or depth, values for tile layers. It would be more like this

>    map.Draw(camera);
>    ...
>    spriteBatch.Begin(...);
>    spriteBatch.Draw(...);
>    spriteBatch.Draw(...);
>    spriteBatch.End();

Okay, I'm fairly convinced this doesn't work.

I spent some time [refactoring](https://github.com/craftworkgames/MonoGame.Extended/commit/6c21bc1433a68007b06ba4921de75492f336c61c) the TiledMap renderer to always render layers between 0.0 and -1.0 depth to match what the `SpriteBatch` does. Then I rendered some sprites using `SpriteSortMode.FrontToBack` and set some depth values.

No matter what I do I can't get the sprites to render in between the layers. It seems that whatever is happening in the `SpriteBatch` is forcing the sprites to render on top of the map regardless of the specified depth value.

**Comment by guFalcon** [comment]
AFAIK the SpriteBatch.End() call draws the batch. No matter what you do. That's why I wrote that they would have to share the batch that they want to draw their sprites in.

> First layer was ground. Second layer was obstacles. Then the developer would draw all sprites within the same batch as the second layer. Third layer was roofs, etc...

The overlapping-mode for two or more batches that certainly works for `Deferred` is just allowing to draw to the various instances. The End()-call still draws the respective batch.

And, I agree, it's a pain in the ass to get such behavior as sharing a batch and sorting based on a float into an API in a nice way.
But you two are the pros on those subjects ;)

BTW... (joking) Have you (Dylan and Lithium) ever considered using Skype?
I've always wondered how you two find time to write such lengthy messages to each other. And you seem to write at the same time as well. At least I usually get all your updates in a batch.
On the other hand, it's like reading the exchange of letters between Schiller and Goethe :) .
Very interesting to read and I've learned a lot from them.
(serious) Have you ever considered putting them up on a blog? Really. I mean that.
At least tell me you've got a link-collection so that I don't have to go over this github-project for the tons of useful links you two have been throwing at each other for the last two month only.
They really would be a great help for other to-be-game-programmers if categorized nicely.
They helped me a lot for sure.
If you don't have such a thing maybe you should consider making a page in the Wiki or something. I certainly would link that page everywhere I can.
Keep up the perfect work and discussions and thank you.

**Comment by craftworkgames** [comment]
Yep. I've just confirmed it to make sure I wasn't going crazy. If it where true that you could set the depth of sprites between batches, these two images would be the same.

![image](https://cloud.githubusercontent.com/assets/3201643/17664201/cd295786-6335-11e6-84e9-904fc43cee08.png)

Clearly, they are not.

Here's the code I used.

    _spriteBatch.Begin(sortMode: SpriteSortMode.BackToFront);
    _spriteBatch.Draw(_logo, position: new Vector2(0, 0), color: Color.Green, layerDepth: 0.5f);
    _spriteBatch.Draw(_logo, position: new Vector2(50, 50), color: Color.Blue, layerDepth: 0.25f);
    _spriteBatch.Draw(_logo, position: new Vector2(100, 0), color: Color.Black, layerDepth: 0.75f);
    _spriteBatch.End();


    _spriteBatch.Begin(sortMode: SpriteSortMode.BackToFront);
    _spriteBatch.Draw(_logo, position: new Vector2(300, 0), color: Color.Green, layerDepth: 0.5f);
    _spriteBatch.End();

    _spriteBatch.Begin(sortMode: SpriteSortMode.BackToFront);
    _spriteBatch.Draw(_logo, position: new Vector2(350, 50), color: Color.Blue, layerDepth: 0.25f);
    _spriteBatch.End();

    _spriteBatch.Begin(sortMode: SpriteSortMode.BackToFront);
    _spriteBatch.Draw(_logo, position: new Vector2(400, 0), color: Color.Black, layerDepth: 0.75f);
    _spriteBatch.End();

@guFalcon Skype, no. But we often chat in the gitter chat room. As for blogging, I used to do that but it can be very time consuming. For a while it was a good way to get people interested in the library, but there are more effective ways to do that now, like answering peoples questions on forums and stackoverflow.

**Comment by LithiumToast** [comment]
@craftworkgames 

You need to enable to depth buffer.

```
            GraphicsDevice.Clear(Color.Gray);

            _spriteBatch.Begin(sortMode: SpriteSortMode.BackToFront, depthStencilState: DepthStencilState.Default);
            _spriteBatch.Draw(_logo, position: new Vector2(0, 0), color: Color.Green, layerDepth: 0.5f);
            _spriteBatch.Draw(_logo, position: new Vector2(50, 50), color: Color.Blue, layerDepth: 0.25f);
            _spriteBatch.Draw(_logo, position: new Vector2(100, 0), color: Color.Black, layerDepth: 0.75f);
            _spriteBatch.End();

            _spriteBatch.Begin(sortMode: SpriteSortMode.BackToFront, depthStencilState: DepthStencilState.Default);
            _spriteBatch.Draw(_logo, position: new Vector2(300, 0), color: Color.Green, layerDepth: 0.5f);
            _spriteBatch.End();

            _spriteBatch.Begin(sortMode: SpriteSortMode.BackToFront, depthStencilState: DepthStencilState.Default);
            _spriteBatch.Draw(_logo, position: new Vector2(350, 50), color: Color.Blue, layerDepth: 0.25f);
            _spriteBatch.End();

            _spriteBatch.Begin(sortMode: SpriteSortMode.BackToFront, depthStencilState: DepthStencilState.Default);
            _spriteBatch.Draw(_logo, position: new Vector2(400, 0), color: Color.Black, layerDepth: 0.75f);
            _spriteBatch.End();
```

Output: 

<img width="275" alt="screen shot 2016-08-15 at 8 02 07 am" src="https://cloud.githubusercontent.com/assets/519592/17668542/9f07fc58-62be-11e6-8f60-5ef805a09e94.png">

But the one on the right doesn't look right does it? That's because the blue and green logo have transparent pixels set in the depth buffer which have a closer Z value. When it's time to render the black logo, the depth buffer is checked which already has a pixels with a closer Z value so part of the back logo is not rendered correctly. This is exactly what Shawn was describing in his blog post.

Let's just double-check that using painter's algorithm for back-to-front solves the artifacts. By default `AlphaBlend`

```
            _spriteBatch.Begin(sortMode: SpriteSortMode.BackToFront, depthStencilState: DepthStencilState.Default);
            _spriteBatch.Draw(_logo, position: new Vector2(0, 0), color: Color.Green, layerDepth: 0.5f);
            _spriteBatch.Draw(_logo, position: new Vector2(50, 50), color: Color.Blue, layerDepth: 0.25f);
            _spriteBatch.Draw(_logo, position: new Vector2(100, 0), color: Color.Black, layerDepth: 0.75f);
            _spriteBatch.End();

            _spriteBatch.Begin(sortMode: SpriteSortMode.BackToFront, depthStencilState: DepthStencilState.Default);
            _spriteBatch.Draw(_logo, position: new Vector2(400, 0), color: Color.Black, layerDepth: 0.75f);
            _spriteBatch.End();

            _spriteBatch.Begin(sortMode: SpriteSortMode.BackToFront, depthStencilState: DepthStencilState.Default);
            _spriteBatch.Draw(_logo, position: new Vector2(300, 0), color: Color.Green, layerDepth: 0.5f);
            _spriteBatch.End();

            _spriteBatch.Begin(sortMode: SpriteSortMode.BackToFront, depthStencilState: DepthStencilState.Default);
            _spriteBatch.Draw(_logo, position: new Vector2(350, 50), color: Color.Blue, layerDepth: 0.25f);
            _spriteBatch.End();
```
Yup. Output:

<img width="268" alt="screen shot 2016-08-15 at 8 09 40 am" src="https://cloud.githubusercontent.com/assets/519592/17668755/ace472ce-62bf-11e6-8c8e-f2a589490c8b.png">

Alright. How about using Alpha Test like Shawn mentioned to solve the artifacts?

```
            _logo = Content.Load<Texture2D>("logo-square-128");
            _spriteBatch = new SpriteBatch(GraphicsDevice);
            _alphaTestEffect = new AlphaTestEffect(GraphicsDevice);
            var viewport = GraphicsDevice.Viewport;
            _alphaTestEffect.Projection = Matrix.CreateTranslation(-0.5f, -0.5f, 0) * Matrix.CreateOrthographicOffCenter(0, viewport.Width, viewport.Height, 0, 0, -1);
            _alphaTestEffect.VertexColorEnabled = true;
            _alphaTestEffect.ReferenceAlpha = 249;

...

            _spriteBatch.Begin(sortMode: SpriteSortMode.BackToFront, depthStencilState: DepthStencilState.Default, effect: _alphaTestEffect);
            _spriteBatch.Draw(_logo, position: new Vector2(0, 0), color: Color.Green, layerDepth: 0.5f);
            _spriteBatch.Draw(_logo, position: new Vector2(50, 50), color: Color.Blue, layerDepth: 0.25f);
            _spriteBatch.Draw(_logo, position: new Vector2(100, 0), color: Color.Black, layerDepth: 0.75f);
            _spriteBatch.End();

            _spriteBatch.Begin(sortMode: SpriteSortMode.BackToFront, depthStencilState: DepthStencilState.Default, effect: _alphaTestEffect);
            _spriteBatch.Draw(_logo, position: new Vector2(300, 0), color: Color.Green, layerDepth: 0.5f);
            _spriteBatch.End();

            _spriteBatch.Begin(sortMode: SpriteSortMode.BackToFront, depthStencilState: DepthStencilState.Default, effect: _alphaTestEffect);
            _spriteBatch.Draw(_logo, position: new Vector2(350, 50), color: Color.Blue, layerDepth: 0.25f);
            _spriteBatch.End();

            _spriteBatch.Begin(sortMode: SpriteSortMode.BackToFront, depthStencilState: DepthStencilState.Default, effect: _alphaTestEffect);
            _spriteBatch.Draw(_logo, position: new Vector2(400, 0), color: Color.Black, layerDepth: 0.75f);
            _spriteBatch.End();
```

Boom. Output:

<img width="276" alt="screen shot 2016-08-15 at 8 28 01 am" src="https://cloud.githubusercontent.com/assets/519592/17669351/36e7c7b2-62c2-11e6-9e75-b7a7c361937d.png">

**Comment by craftworkgames** [comment]
Incredible :smile: 

One thing that made me laugh is that `DepthStencilState.Default` is not actually the default value. Heh.

Anyways, I hacked in your changes and got some sprites rendering between Tiled layers. Boom! :wink: 

![image](https://cloud.githubusercontent.com/assets/3201643/17681860/73c113c2-638b-11e6-9dbd-f540a787ed79.png)

So we have one possible solution. I'd still like the ability to render each layer individually so that people can use the painter's algorithm if they want.

I also noticed a fair amount of old code floating around that can be cleaned up.

I'm thinking we should have some sensible default depths when rendering layers and tiles in isometric maps but also provide a way to override these defaults. I also think it might be worth considering what other parameters should be passed into the `Draw` method if any.

**Comment by TheUltimateHose** [comment]
That solution looks good for platformers, but I think it wouldn't work that well for RPG style maps, as an entity in a RPG can stand both behind and in front objects of a layer. Because of that, I created my own "little" work-around which can be found in **[this gist](https://gist.github.com/TheUltimateHose/c91d17f96a13a4c409b1892d305bf612)**. I hope the code is easy enough to understand, but the idea behind this is that tiles and entities are basically equal when it comes to rendering, with the "layer" of such objects in an RPG style map being defined by their baseline (y + height).

But the code is by far not perfect: It still has the gap issue, as I didn't come around to understanding that primitive rendering stuff, but I came up with a work-around for that too: I round the position of my camera which works fine for me.

A complete call to the renderer would look like this:
```csharp
List<TiledTileLayer> ground = new List<TiledTileLayer>();
ground.Add(_map.GetLayer<TiledTileLayer>("Layer 1"));

List<TiledTileLayer> others = new List<TiledTileLayer>();
others.Add(_map.GetLayer<TiledTileLayer>("Layer 2"));

List<Renderable> entities = new List<Renderable>();
entities.Add(new Renderable(_player._sprite));

_map.Draw(_spriteBatch, ground, others, entities);
```

An example of that renderer in Action:
- Player standing in front of an object
![before](https://cloud.githubusercontent.com/assets/10011726/17833444/cb7fac28-671d-11e6-8df6-5a593a440604.png)
- Player standing behind the same object
![behind](https://cloud.githubusercontent.com/assets/10011726/17833453/00c06c2e-671e-11e6-9c83-8b615668d243.png)

**Comment by LithiumToast** [comment]
@TheUltimateHose > as an entity in a RPG can stand both behind and in front objects of a layer.

That's why I suggested to promote group of tiles to entities in some cases; in this case the fire hydrant. Then it's just sorting the entities for depth. With this idea, tiles would be the static things in the scene like the road and buildings while the player, cars, fire hydrant, garbage can, even the doors if they should animate to open, would be the dynamic things or entities.

> the idea behind this is that tiles and entities are basically equal when it comes to rendering

Unfortunately that's no longer possible as tiles no longer use `SpriteBatch` to render. The new tiled renderer creates the map geometry _once_ ahead of time and stores it in a _static_ `VertexBuffer`. The old way of rendering a tiled map used `SpriteBatch` which the geometry is created _once every frame_ and stored in a _dynamic_ `VertexBuffer`. One of the majors differences is the CPU no longer has process every tile every frame with the new tiled renderer, increasing performance, especially for extremely large maps. The other major difference is that the geometry is only uploaded once to the GPU instead of every frame, which also has a performance boost. Rendering using the new tiled renderer just tells the GPU to process the vertices already in video memory from a starting location to an ending location in the buffer for rendering.

**Comment by craftworkgames** [comment]
Let's just keep in mind that nothing is set in stone at this stage. Our goal here is to make it easier to make games and if the approach we are currently taking doesn't allow that, we'll change it.

The old approach and the new approach both have pros and cons. The purpose of this discussion is to iron out the problems with the new approach so that we can get the best of both worlds. The number one goal is to make games, but we can't do that if rendering the map chews up all the memory and CPU.

I have no doubt that there's a reasonable hybrid approach rendering some of it as static geometry and some of it as dynamic geometry. Make no mistake, we will continue to make improvements until the library is suitable for making games. Unfortunately, these things do take some time, especially in an open source project.

**Comment by rafaelalmeidatk** [comment]
I have an idea: the problem here is primarly focused on the Z values. We need something to **sort the Z values** before drawing.

To begin, we can split the map in few _layers_:

- Ground Layer = Layer 1
- Entities Layer = Layer 2
- Details Layer = Layer 3

Ok, so the idea here is basically that: the layers are always one on top of other, **they are on different Z values and never cross against each other**.

Now lets talk about each layer we created individually:

Ground Layer
----
This one is imported from the TileD. We have tiles and they are always behind everything. All the components of this layer have the same Z value.

Entities Layer
----
Also imported from the TileD, but now we have tiles and `Sprites` that we will add dinamically. All the `Sprite`s have a Z value. We can link the `Sprite`s to the map like that (I am writing the code that comes in my mind, of curse we will name the things better):

    Sprite player = new Sprite();
    Sprite npc = new Sprite();

    entitiesLayer = map.GetLayer("Entities");
    entitiesLayer.linkSprite(player);
    entitiesLayer.linkSprite(npc);
    entitiesLayer.setDynamicZ(true);

Now our `SpritesMapLayer` will assign the correct Z values to the `Sprites` based on their Y values, and this value will be between the Layer 1 and Layer 3. What we have to do now, is know **how _update_ the value of the vertices with the new Z values**. If we accomplish it, we will have dynamic sprites and tiles sorting.

Details Layer
-----
Same to the Ground Layer, but this one will go on top of everything.

---

I don't think that will solve all the issues, I am just trying to contribute with the vision of someone that is really using the library to create a real game and in a way that don't is soo complicated, because I don't think that the complex way is the best to we go, what the purpose of our library if it is too complex to the others understand or even use?

**Comment by LithiumToast** [comment]
>Now our SpritesMapLayer will assign the correct Z values to the Sprites based on their Y values, and this value will be between the Layer 1 and Layer 3. What we have to do now, is know how update the value of the vertices with the new Z values. If we accomplish it, we will have dynamic sprites and tiles sorting. We can draw the sprites using the SpriteBatch and depth buffer directly on the map renderer, this will make all the drawing be transparent to the developer.

What if people don't want to sort by Y values but by something else? What if people don't want to use depth buffering at all, but instead use the painter's algorithm? I don't even see the point of an entities layer when that's basically what a `SpriteBatch`'s `Begin`/`End` pair is, which offers deferred sorting of draw calls by depth values for correct output with the painter's algorithm.

**Comment by McSora** [comment]
@LithiumToast 

> That's why I suggested to promote group of tiles to entities for some games;

It's been a while, got caught up a little, but found some time to go through the code and ... I came to the conclusion that your suggestion would be the best solution, at least in my case (i guess i am not the only one who's beginning to understand that there obviosly are a million of different scenarios out there :smile:)

So i wanted to exchange some thoughts on the workflow:
The Tiled Editor features the Tile-Layer, the Object-Layer (ObjectGroup as called in the API) and the Image Layer.
- I believe the Image Layer is only for one Image in the Background so that's that.
- The Tile-Layer in your concept would only be used for static tiles like ... grass tiles... water tiles... stuff a character only walks on or stuff that is only above all.
- and the object groups would be the place where you'd put the entities, am i right?

So i looked around for a bit (the tiled editor seems so small but has so much to offer) and created an object-layer. in here you would be able to place tiles pixelperfect.
So, one could create an entity-tileset consisting of only the entities (the group of tiles that would build the whole thing, like a large tree, or eben a one-tile rock one could stand behind of) and these can be placed as objects.

After one painted his static tiles on a custom number of tile-layers, maybe set a background image, showing the flames of everlasting purgatory, and placed all the rocks, trees and houses as tiles on the object-layer, the Content.Load<TiledMap> enters the stage.

In the resulting object TileLayers stay TileLayers, the ImageLayer stays the ImageLayer and the Objects go into the ObjectGroup. 
I saw that the object hold their type (in this case they would be Type.Tile) and a GID. Didnt have the time to check if this was the TileID if the type was Type.Tile, but the nullable int kind of gave me the idea.
So you would have the object with position and all, and the visual reference to the TileID (maybe).

Then I searched for the point where objects of the objectGroups are drawn but either I am blind, for which i then would be really sorry :smile:, or they aren't drawn at all. 
The latter would mean that one would in the end have to draw the objects on his own of course, looping through the objectGroup (and maybe other entities not defined int he TMX), holding a reference to the (now public) tilesets of the map, resolving the GID of the object to get the wanted tile and then drawing it in a batch call, sorted the way one likes.

Hope i didn't miss anything, but... would this in any way sum up your suggested concept? :)

**Comment by LithiumToast** [comment]
@McSora If I were doing it, I would parse the map for objects once it is loaded with `Content.Load<TiledMap>` to create the entities/sprites which would be held in an array or list somewhere. Then every frame to draw the entities just loop through the array/list of entities with a `SpriteBatch` to render them. With the parsing of tiled objects it's easier to ignore or exclude objects in case of quests or something like items on the map which can only be picked up once for example.

**Comment by McSora** [comment]
@LithiumToast 
Thanks for the quick answer :D

Then i guess i got an idea of how to do this... the exclusion into a list of entities outside of the map would also be the first thing i'd do for there will be dynamically created entities that would have to be sorted the same way ;D

I kind of just wanted to check the Tiled-Editor workflow xD If the object layer and the Tile-Type Objects would be the thing you would go for as well.
And maybe even remember this whole scenario and workflow for a **future demo game**.

We all just want to use the api how it is supposed to be used and get that nice warm feeling in the heart that i got when i saw my tiled-map getting rendered the first time with only 3 lines of code written :Smile:

**Comment by Taikatou** [comment]
I have a working example here https://github.com/Taikatou/PureGame not perfect but wanted to see how I would make an RPG with TiledMap

**Comment by craftworkgames** [comment]
Thanks @Taikatou. I tried to compile PureGame and ran into a bunch of issues. Eventually I got it to compile but it's not showing much.

In any case I do appreciate real world examples of usage. I think that's the best way to make sure the library is doing what it's supposed to do.

**Comment by Taikatou** [comment]
@craftworkgames Wow did not expect you to open project, it is currently broken (Forgot to commit fix) I really like what we currently have and think TiledMap should be an empty vessel everyone can build anything with.

**Comment by rafaelalmeidatk** [comment]
@LithiumToast you are right, my solution wasn't generic.

> That's why I suggested to promote group of tiles to entities for some games

This is the way to go. I did a PR ( #257 ) that uses this idea, the good news are that the solution is way more generic now. I'll try to create a new demo to top-down games.

**Comment by craftworkgames** [comment]
Closing this issue. We'll continue to discuss the evolution of the new renderer in #268

**Comment by kerams** [comment]
It's now been a couple of years and the library has evolved. Is there a solution to the problem @0xDiddi described in the latest version?
